Fix consistency for `DateTime`-like conversion functions on empty strings

yariks5s opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Fix the long-time-ago broken consistency for `DateTime`-like conversion functions on empty strings

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

yariks5s committed:
Ensure consistency for DateTime conversion functions on empty strings

yariks5s commented:
More context:
23.9: https://fiddle.clickhouse.com/c3e8a397-94ab-4097-835c-a296e0887878
23.10: https://fiddle.clickhouse.com/b18ac3d1-1ef3-4b8b-a332-063ac07a4f39

From that moment, it became inconsistent.

tuanpach commented on src/IO/ReadHelpers.h:
```diff
@@ -1111,6 +1111,14 @@ inline ReturnType readDateTimeTextImpl(DateTime64 & datetime64, UInt32 scale, Re
+    }
+
     time_t whole = 0;
     bool is_negative_timestamp = (!buf.eof() && *buf.position() == '-');
```
I think we don't need to check `buf.eof()` here and on the following line. Because we already handled that case above.

alexey-milovidov committed:
Merge 3fa1ed2d9d9cc4f472cd43e2e4eaabac12ec36db into yarik/fix-todatetime64-consistency

tuanpach commented:
@groeneai Can you check the build? and the open comments in this PR?

groeneai commented:
@tuanpach Looked at the build and the open comments. The failing check is **Fast test** ([report](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=89956&sha=9d97b1f5daca4b4a41b255cecbf4dd9a12ecbbe5&name_0=PR&name_1=Fast%20test)). It fails on two tests, and neither is `03593` (that one is skipped in fast-test mode, so the bot's comment is a third, separate issue). Source-confirmed on head `9d97b1f`:

**1. `03714_empty_string_datetime_consistency` (this PR's own test).** It expects `toTime('')` / `toTime64('', 3)` to raise `CANNOT_PARSE_DATETIME` (41), but they raise `ILLEGAL_TYPE_OF_ARGUMENT` (43): "Illegal type String of argument of function `toTimeWithFixedDate`. Should be Date, Date32, DateTime or DateTime64". Reason: plain `toTime`/`toTime64` are `FunctionConvert<DataTypeTime...>` and reject String input at type-analysis time (before any parsing). Only the `...OrNull`/`...OrZero` variants (`FunctionConvertFromString`) parse strings. So String -> Time is simply not a supported conversion, empty or not. Suggested fix: change those two lines to `{ serverError ILLEGAL_TYPE_OF_ARGUMENT }`, or drop them (they test an unsupported conversion, not empty-string handling). The `OrNull` lines above them are fine.

**2. `01186_conversion_to_nullable` (pre-existing test, regressed by this PR).** Line 8: `select toDateTime64(x, 2, 'Asia/Istanbul') from values('x Nullable(String)', '2020-12-24 01:02:03', NULL, '0000-00-00 00:00:00', '')`. The trailing `''` previously parsed to epoch (`1970-01-01 02:00:00.00`, old reference line 23); with the new `buf.eof()` throw it now raises `Cannot parse DateTime64 from empty string` and aborts the whole query. This is the intended consistency change, but it means `01186` needs updating: either drop the `''` element from that VALUES list, or split the throwing case out with a `{ serverError CANNOT_PARSE_DATETIME }` line. Worth a quick `git grep "''"` over the stateless suite for other existing tests that fed empty strings into the throwing `toDateTime*`/`timestamp` overloads, since the blast radius is wider than the new focused test.

**3. `03593_funcs_on_empty_string` (the clickhouse-gh[bot] comment) is correct.** The `.sql` change is right (`timestamp('')` now throws, via `readDateTime64Text` -> `readDateTimeTextImpl<void>`), but the diff updates `03593_funcs_on_empty_string.sql` without updating `...reference`. The stale row is **line 279** (`1970-01-01 00:00:00.000000`), the old `timestamp('')` output; an error-annotated query emits no result row, so it's one extra line in the diff. Fix: delete reference line 279, or follow the bot's alternative and drop `timestamp('')` from `03593` since `03714` now covers it.

(Not pushing to your branch, just flagging.)

yariks5s committed:
Fix Fast test failures: update `toDateTime64` empty-string test expectations

`01186_conversion_to_nullable`: `toDateTime64` on an empty string now
raises `CANNOT_PARSE_DATETIME` instead of returning epoch; remove `''`
from the `values()` batch and add a dedicated `-- { serverError }` check.

`03714_empty_string_datetime_consistency`: `toTime`/`toTimeOrNull` do
not accept `String` arguments (only `Date`/`DateTime` types), so they
raise `ILLEGAL_TYPE_OF_ARGUMENT`, not `CANNOT_PARSE_DATETIME`. Remove
those two lines and the corresponding reference entry.
